Containarize py api

IvanObreshkov opened the pull request:
* Added Dockerfile
* Made the server start with multiple workers if we are in PROD

asynchroza commented on services/py-api/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+WORKDIR /app
+
+COPY src ./src
+COPY .env ./
```
Wouldn't these be injected by the compose configuration? I honestly believe that you shouldn't depend on the `.env` file here. If you're spinning up just this image, pass them inline.

IvanObreshkov replied:
I think it will be passed by the docker-compose. This PR is only for the docker image itself. I thought of making the docker-compose in another, once we clear this one.

asynchroza replied:
Yes, that's what I thought as well. I believe you should get rid of this `COPY` then.

asynchroza commented on services/py-api/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+
+COPY pyproject.toml poetry.lock README.md ./
+
+RUN poetry install --no-root && rm -rf $POETRY_CACHE_DIR
```
Do you have dev dependencies? This will install everything.

IvanObreshkov replied:
Yeah we do. Good catch, I forgot to exclude them.

asynchroza commented on services/py-api/.env.sample:
```diff
@@ -1,4 +1,5 @@
 PORT=8080
-ADDRESS=localhost
+# means "all IPv4 addresses on the local machine". This is needed for Docker.
+ADDRESS="0.0.0.0"
```
Why is this needed for docker?

IvanObreshkov replied:
If I run the image with `-p 127.0.0.1:8080:8080`, I am not able to access the API when pinging it on `https://localhost/api/v3/ping`. It says connection was closed.

asynchroza replied:
Okay. I read a bit more about it. That's fine security wise.

asynchroza replied:
Btw, I'm seeing you have `v3` in your urls. Does it make sense to continue the versioning if both the other two APIs are deprecated and discarded?

iobreshkov-qb replied:
> Btw, I'm seeing you have `v3` in your urls. Does it make sense to continue the versioning if both the other two APIs are deprecated and discarded?

I left it on purpose, as a jem :D. To remind us that we had 2 previous APIs.

Isak-Bego replied:
> Why is this needed for docker?

I confirm this because when I was setting the dev container I had to change the address to 0.0.0.0 for the localhost to work.

asynchroza commented on services/py-api/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+RUN apt-get update && apt-get install -y --no-install-recommends curl && rm -rf /var/lib/apt/lists/*
+
+ENV VIRTUAL_ENV=/app/.venv \
+    PATH="/app/.venv/bin:$PATH"
```
I would suggest that you add a comment explaining what you have done here and why this is required. 
On such builds I would usually opt out of using a virtual environment and directly pull my packages into `site-packages`.

asynchroza commented on services/py-api/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+EXPOSE 8080
+
+# As we are using self-signed certs we don't check the validity of the certs with -k
+HEALTHCHECK --interval=1m --timeout=10s CMD curl -k -f https://localhost:6969/email-service/v1/ping
```
This might be out of scope but what I would suggest you do here is:

- Have the domain as an environment variable so that production containers run the check against the proper url and catch ssl issues.
- If you're opting into running self-signed certs, install them into the container so that you don't have to skip verification.

IvanObreshkov replied:
I like the idea with the env variable for the domain. Apparently i have forgotten to copy the certs folder. Nevertheless, as this will be running only on DEV and PROD environmentsI don't think we need the self signed certs at all. What do you think? Because if we do have them the healthcheck will have to choose to verify the certs or not to verify them, depending on the domain. If it's localhost we don't verify them, if it's not we should verify them.

I am a bit concerned if the check towards https://thehub-aubg.com/api/v3/ping will work, as last time I did that I had some issues with it. Have to try it.

asynchroza replied:
I think the idea with self-signed certs is good. Something which was an issue with the previous configuration was that it was incredibly hard to debug in local environments. You couldn't just spin up everything from the get go and verify if requests are propagated correctly between containers.

asynchroza commented on services/py-api/src/server/config/server_config.py:
```diff
@@ -41,9 +41,14 @@ def start() -> None:
+            # https://docs.gunicorn.org/en/stable/design.html#how-many-workers
+            # As cpu_count could return None we use 0 instead, as 2 * None would produce an error
+            # Also "workers" flag is ignored when reloading is enabled.
+            workers=(2 * (cpu_count() or 0) + 1) if server_config.ENV == "DEV" else None,
```
That's cool. I didn't know that there's such an option. How are requests load balanced?

Also, why is this enabled only for DEV? I think it would make sense to have it on PROD as well, no?

IvanObreshkov replied:
I was testing something and forgot it. Will fix it tomorrow.

IvanObreshkov commented:
@asynchroza Could you review it again? I added more docs and rewrote the `docker-compose.yaml`. We have a deployed version, which you can check at https://dev.thehub-aubg.com:8080/api/v3/ping

The health check works and is pinging on the url above.
<img width="1295" alt="image" src="https://github.com/user-attachments/assets/5fccd923-1e6b-4aee-b1b8-d4838743ccae">

I will also add https://github.com/willfarrell/docker-autoheal in another PR once we merge this one

IvanObreshkov commented:
> @asynchroza Could you review it again and resolve the comments? I added more docs and rewrote the `docker-compose.yaml`. We have a deployed version, which you can check at https://dev.thehub-aubg.com:8080/api/v3/ping
> 
> The health check works and is pinging on the url above. <img alt="image" width="1295" src="https://private-user-images.githubusercontent.com/38375079/376776331-5fccd923-1e6b-4aee-b1b8-d4838743ccae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwMjM2NzksIm5iZiI6MTcyOTAyMzM3OSwicGF0aCI6Ii8zODM3NTA3OS8zNzY3NzYzMzEtNWZjY2Q5MjMtMWU2Yi00YWVlLWIxYjgtZDQ4Mzg3NDNjY2FlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE1VDIwMTYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhNGMyYWI0MzFhNjc4MWZmMmIzMTI1MzY2Y2U5YTVmOThlOGE3MDk0NjBlZDFjYTNhYTQzNzVhOTM1ODJjMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.x11zgQu7yulKTikIrNpR3nZ626IEt3SelidIlSlMv0k">
> 
> I will also add https://github.com/willfarrell/docker-autoheal in another PR once we merge this one

Btw in the logs sometimes I see this, we should investigate:
`py-api  | INFO:     172.20.0.1:54710 - "GET /api/v3/ping HTTP/1.1" 200 OK
py-api  | Traceback (most recent call last):
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 1261, in connect
py-api  |     sock = _configured_socket(self.address, self.opts)
py-api  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 862, in _configured_socket
py-api  |     sock = _create_connection(address, options)
py-api  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 815, in _create_connection
py-api  |     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
py-api  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
py-api  |   File "/usr/local/lib/python3.12/socket.py", line 976, in getaddrinfo
py-api  |     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
py-api  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
py-api  | socket.gaierror: [Errno -2] Name or service not known
py-api  | 
py-api  | The above exception was the direct cause of the following exception:
py-api  | 
py-api  | Traceback (most recent call last):
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/mongo_client.py", line 2064, in _process_periodic_tasks
py-api  |     self._topology.update_pool()
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/topology.py", line 676, in update_pool
py-api  |     server.pool.remove_stale_sockets(generation)
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 1214, in remove_stale_sockets
py-api  |     conn = self.connect()
py-api  |            ^^^^^^^^^^^^^^
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 1281, in connect
py-api  |     _raise_connection_failure(self.address, error, timeout_details=details)
py-api  |   File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 208, in _raise_connection_failure
py-api  |     raise AutoReconnect(msg) from error
py-api  | pymongo.errors.AutoReconnect: thehubwebsite-shard-00-00.h9aqj.mongodb.net:27017: [Errno -2] Name or service not known (configured timeouts: connectTimeoutMS: 20000.0ms)
`

asynchroza commented on services/py-api/src/server/config/server_config.py:
```diff
@@ -28,19 +35,18 @@ def start() -> None:
-            log_config=get_uvicorn_logger(server_config.ENV),
+            # Reloading is intended only for local development. Simulating a PROD/DEV env will result in multiple
+            # workers spawned
+            reload=server_config.ENV == "LOCAL" and server_config.SIMULATING_PROD_ENV is False,
```
What's the difference between pointing `ENV` to `"PROD"` and setting `SIMULATING_PROD_ENV` to True?

I don’t believe an additional variable is necessary to simulate the environment. For accurate simulation, it’s best to closely resemble the original environment. Adding an extra variable here may simply introduce more factors that could cause divergence from the intended conditions.

If you just want to be able to control the workers, you can extract the count as an environment variable.

`WORKERS_COUNT=0` for local environments

IvanObreshkov replied:
I thought about your proposal. But it doesn't take into consideration our current logging situation. Before the introduction of the `SIMULATING_PROD_ENV`, we sent the logs to a file only if `ENV` is "DEV" or "PROD". But In the `Dokcerfile` we have a check which says if `ENV=LOCAL` don't check the validity of the self-signed certs. In the previous setup (before introducing `SIMULATING_PROD_ENV`), If you change `ENV=DEV`  and run the container locally, the logs will be written to a file, but the certs validation will fail and the container will be marked unhealthy :D. Also, the workers' count is calculated on demand via `(2 x $num_cores) + 1`, and I am not sure if we want to change this. And if reloading is enabled the `workers` flag is ignored, making it a bit of a mess. That's why I introduced the variable.

IvanObreshkov commented:
> > @asynchroza Could you review it again and resolve the comments? I added more docs and rewrote the `docker-compose.yaml`. We have a deployed version, which you can check at https://dev.thehub-aubg.com:8080/api/v3/ping
> > The health check works and is pinging on the url above. <img alt="image" width="1295" src="https://private-user-images.githubusercontent.com/38375079/376776331-5fccd923-1e6b-4aee-b1b8-d4838743ccae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwMjM2NzksIm5iZiI6MTcyOTAyMzM3OSwicGF0aCI6Ii8zODM3NTA3OS8zNzY3NzYzMzEtNWZjY2Q5MjMtMWU2Yi00YWVlLWIxYjgtZDQ4Mzg3NDNjY2FlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE1VDIwMTYxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhNGMyYWI0MzFhNjc4MWZmMmIzMTI1MzY2Y2U5YTVmOThlOGE3MDk0NjBlZDFjYTNhYTQzNzVhOTM1ODJjMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.x11zgQu7yulKTikIrNpR3nZ626IEt3SelidIlSlMv0k">
> > I will also add https://github.com/willfarrell/docker-autoheal in another PR once we merge this one
> 
> Btw in the logs sometimes I see this, we should investigate: `py-api | INFO: 172.20.0.1:54710 - "GET /api/v3/ping HTTP/1.1" 200 OK py-api | Traceback (most recent call last): py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 1261, in connect py-api | sock = _configured_socket(self.address, self.opts) py-api | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 862, in _configured_socket py-api | sock = _create_connection(address, options) py-api | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 815, in _create_connection py-api | for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM): py-api | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ py-api | File "/usr/local/lib/python3.12/socket.py", line 976, in getaddrinfo py-api | for res in _socket.getaddrinfo(host, port, family, type, proto, flags): py-api | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ py-api | socket.gaierror: [Errno -2] Name or service not known py-api | py-api | The above exception was the direct cause of the following exception: py-api | py-api | Traceback (most recent call last): py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/mongo_client.py", line 2064, in _process_periodic_tasks py-api | self._topology.update_pool() py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/topology.py", line 676, in update_pool py-api | server.pool.remove_stale_sockets(generation) py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 1214, in remove_stale_sockets py-api | conn = self.connect() py-api | ^^^^^^^^^^^^^^ py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 1281, in connect py-api | _raise_connection_failure(self.address, error, timeout_details=details) py-api | File "/app/.venv/lib/python3.12/site-packages/pymongo/synchronous/pool.py", line 208, in _raise_connection_failure py-api | raise AutoReconnect(msg) from error py-api | pymongo.errors.AutoReconnect: thehubwebsite-shard-00-00.h9aqj.mongodb.net:27017: [Errno -2] Name or service not known (configured timeouts: connectTimeoutMS: 20000.0ms) `
> 
> This exception does not mark the container unhealthy and doesn't make it restart.
> 
> Maybe a retry logic with exponential backoff will help.
> 
> This is also an option as we are using multiple server workers: https://pymongo.readthedocs.io/en/stable/faq.html#multiprocessing
> 
> Or this one: https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html ![Screenshot_20241016_085505_Chrome.jpg](https://private-user-images.githubusercontent.com/38375079/376909144-afd64ebc-578c-488d-9ee4-8d844d03b04f.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MzA2MzEsIm5iZiI6MTcyOTYzMDMzMSwicGF0aCI6Ii8zODM3NTA3OS8zNzY5MDkxNDQtYWZkNjRlYmMtNTc4Yy00ODhkLTllZTQtOGQ4NDRkMDNiMDRmLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIyVDIwNTIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkyM2E2YzcwMjA1NTcwOWQzYTA3Y2M3NjAyNTQ4Mzk2NzhiM2NhZWM3MWZmYTQ1YWE2NGFlOGVkNmQ2YmRjOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pn6-MyUnMEmZw-ytOxFdq1POPvZ7vrKjUCl4IxIlEVg)
> 
> Current # of conn is 61 which is way before the max of 250 <img alt="image" width="871" src="https://private-user-images.githubusercontent.com/38375079/376792737-e180c18a-ca10-4ec9-8442-d2cffebc358e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MzA2MzEsIm5iZiI6MTcyOTYzMDMzMSwicGF0aCI6Ii8zODM3NTA3OS8zNzY3OTI3MzctZTE4MGMxOGEtY2ExMC00ZWM5LTg0NDItZDJjZmZlYmMzNThlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIyVDIwNTIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwN2FjYzI0NmNiN2NhZTk0MWUzZmQ1NWYyOThiZjg4ZjEwYzZiNjdkMzE0OTE5NGY3Yzk3ZTAwNjM4N2ZkN2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sWIx3xdy8EG5HjyEu99teakzWVgu89BIoBLDTI72ZBU">

Apparently, this issue has been resolved, maybe by some of my previous commits. I have been monitoring the machine for around 15 min and no issue so far. I will check tomorrow, but hopefully we are good.

IvanObreshkov commented on services/py-api/src/server/logger/logger_factory.py:
```diff
@@ -88,13 +90,22 @@ def configure_app_logger(env: str) -> None:
+            # For DEV and PROD as these are VMs we save the logs to a logfile, so that we can check them later
+            # For LOCAL and TEST env we print the logs directly to the stdout
+            # FIXME: This could create an issue as the uvicorn logger will eventually create a server.log.1 file but
+            #  these logs will continue to be written to the old server.log file
```
@asynchroza @Isak-Bego I'm not sure how to fix this from the top of my head. Another concern I have is the cluttering of the `server.log` on the DEV machine with health check logs. I have enabled DEBUG mode on line 109 for the DEV machine for easier debugging if there is a problem.
